packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -107,25 +120,39 @@ export class ExecutionEngine {
       })
     );
 
-    while (!isExecutionResult(current)) {
+    while (!isExecutionResultMessage(current)) {
       context.executionState = state.executionStateMap[future.id];
 
       if (isOnChainAction(current)) {
         current = await state.transactionService.onchain(current, {
           libraries,
         });
-      } else if (isOnchainResult(current)) {
+      } else if (isOnChainSuccessMessage(current)) {
         current = (await exectionStrategy.next(current)).value;
+      } else if (isOnChainFailureMessage(current)) {
+        current = this._convertToExecutionFailure(current);
```

Comment:
Failures do not lead to the execution strategy, instead the execution state is updated immediately as a failure.